### PR TITLE
Workflow based tutorial for Tensorflow CNN training on MNIST data

### PR DESCRIPTION
**Pull request by snehal-das**
This PR adds an example for training a Tensorflow CNN to recongnise handwritten digits using an MNIST dataset.

### CODE REVIEW

File: openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "c2352436",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import tensorflow as tf\n",
+    "import tensorflow.python.keras as keras\n",
+    "import matplotlib.pyplot as plt\n",
+    "from keras import backend as K\n",
```
Comment (rahulga1) [inline openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb]: nit, some of the imports are unused and some are reimported.

File: openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "e6c73353",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "flflow = KerasMNISTWorkflow(model, rounds=3, checkpoint=True)\n",
+    "flflow.runtime = local_runtime\n",
+    "flflow.run()"
```
Comment (rahulga1) [inline openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb]: It would be good to add the code to run the final model on the X_test to show the final accuracy.

File: openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb [hunk 3]
```diff
@@ -0,0 +1,320 @@
+{
+ "cells": [
```
Comment (teoparvanov) [inline openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb]: I suggest adding an overview cell, which summarizes the goal and the main steps of the tutorial.

File: openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb
Comment (teoparvanov) [inline openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb:1]: To match the naming pattern of the tutorials, add a `104_` prefix to the filename.

File: openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+    "%pip install git+https://github.com/securefederatedai/openfl.git\n",
+    "%pip install -r workflow_interface_requirements.txt\n",
+    "\n",
+    "#Install Tensorflow and MNIST dataset if not installed\n",
+    "%pip install tensorflow==2.13\n",
+    "\n",
+    "# Uncomment this if running in Google Colab and set USERNAME if running in docker container.\n",
+    "# !pip install -r https://raw.githubusercontent.com/intel/openfl/develop/openfl-tutorials/experimental/workflow_interface_requirements.txt\n",
+    "# import os\n",
+    "# os.environ[\"USERNAME\"] = \"colab\""
```
Comment (teoparvanov) [inline openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb]: Have you tested this in Colab?

File: openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from openfl.experimental.interface import FLSpec, Aggregator, Collaborator\n",
+    "from openfl.experimental.runtime import LocalRuntime\n",
+    "from openfl.experimental.placement import aggregator, collaborator\n",
+    "\n",
+    "agg = Aggregator()\n",
+    "\n",
+    "collaborator_names = [\"C1\", \"C2\"]\n",
```
Comment (teoparvanov) [inline openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb]: As I suggested to Kush also, let's call these after big cities (instead of "C1" and "C2"), to both add personality, but also highlight the geo-distributed aspect of FL!

**Review by teoparvanov — APPROVED** [review]
LGTM, with a couple of minor comments (in addition to @rahulga1's suggestions)!

File: openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb (hunk 4 above)
Comment (snehal-das) [reply openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb]: No, will test it out. Thanks!

File: openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb (hunk 3 above)
Comment (theakshaypant) [reply openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb]: In addition to the overview, I would also recommend having description before each of the cells explaining the need and functionality of each of the cells. [Ref.](https://github.com/securefederatedai/openfl/blob/develop/openfl-tutorials/experimental/104_Keras_MNIST_with_GPU.ipynb)

**Review by theakshaypant — COMMENTED** [review]
LGTM, just adding on to Teo's suggestion.

File: openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb (hunk 1 above)
Comment (snehal-das) [reply openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb]: Removed redundant packages.

File: openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb (hunk 2 above)
Comment (snehal-das) [reply openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb]: Added loss and accuracy calculations in the end step of the aggregator.

File: openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb
Comment (snehal-das) [reply openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb:1]: Renamed the file to 104_Keras_MNIST_with_CPU.ipynb

File: openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb (hunk 5 above)
Comment (snehal-das) [reply openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb]: Renamed the collaborators.

**Comment by teoparvanov** [comment]
Hi @snehal-das , could you address the DCO job failure, by signing all your commits?
```bash
git commit -s -m "..."
```

File: openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb (hunk 3 above)
Comment (snehal-das) [reply openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb]: Added steps and descriptions to cells where it applies.

File: openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb (hunk 4 above)
Comment (snehal-das) [reply openfl-tutorials/experimental/Keras_MNIST_with_CPU.ipynb]: Tested the changes on Colab.

**Review by rahulga1 — APPROVED** [review]

**Comment by teoparvanov** [comment]
@snehal-das, could you try rebasing in an attempt to fix the GaNDLF test failure in the CI?
